File: src/data/courses/dsa/dynamic-programming/linear-1d-dp/max-sum-non-adjacent/code/java/MaxSumNonAdjacent.java
class Solution {
    public int maxSumNonAdjacent(int[] nums) {
        if (nums.length == 0) return 0;
        int a = Math.max(0, nums[0]), b = nums.length > 1 ? Math.max(a, nums[1]) : a;
        for (int i = 2; i < nums.length; i++) { int c = Math.max(b, a+nums[i]); a=b; b=c; }
        return b;
    }
}

File: src/data/courses/dsa/dynamic-programming/linear-1d-dp/max-sum-non-adjacent/code/cpp/max_sum_non_adjacent.cpp
class Solution {
public:
    int maxSumNonAdjacent(vector<int>& nums) {
        if (nums.empty()) return 0;
        if (nums.size() == 1) return max(0, nums[0]);
        int prev2 = max(0, nums[0]), prev1 = max(prev2, nums[1]);
        for (int i = 2; i < nums.size(); i++) {
            int curr = max(prev1, prev2 + nums[i]);
            prev2 = prev1; prev1 = curr;
        }
        return prev1;
    }
};
